internal/dev_server/ui/src/EnvironmentSelector.tsx

```diff
@@ -17,51 +24,59 @@ export function EnvironmentSelector({
   selectedEnvironment,
   setSelectedEnvironment,
 }: Props) {
-  const [environments, setEnvironments] = useState<Environment[]>([]);
+  const [environments, setEnvironments] = useState<Environment[] | null>(null);
   const [filteredEnvironments, setFilteredEnvironments] = useState<
-    Environment[]
+    Environment[] | null
   >([]);
   const [searchQuery, setSearchQuery] = useState('');
-  const [isLoading, setIsLoading] = useState(true);
+  const [isLoading, setIsLoading] = useState(false);
 
-  useEffect(() => {
-    setIsLoading(true);
-    fetchEnvironments(projectKey)
-      .then((envs) => {
-        setEnvironments(envs);
-        setFilteredEnvironments(envs);
-        if (!selectedEnvironment) {
-          const sourceEnv = envs.find(
-            (env) => env.key === sourceEnvironmentKey,
-          );
-          if (sourceEnv) {
-            setSelectedEnvironment(sourceEnv);
-          } else if (envs.length > 0) {
-            setSelectedEnvironment(envs[0]);
+  const fetchEnvironmentsDebounced = useCallback(
+    debounce((query: string) => {
+      setIsLoading(true);
+      fetchEnvironments(projectKey, query)
+        .then((envs) => {
+          setEnvironments(envs);
+          setFilteredEnvironments(envs);
+          if (!selectedEnvironment) {
+            const sourceEnv = envs.find(
+              (env) => env.key === sourceEnvironmentKey,
+            );
+            if (sourceEnv) {
+              setSelectedEnvironment(sourceEnv);
+            } else if (envs.length > 0) {
+              setSelectedEnvironment(envs[0]);
+            }
           }
-        }
-      })
-      .catch((error) => {
-        console.error('Error fetching environments:', error);
-      })
-      .finally(() => {
-        setIsLoading(false);
-      });
-  }, [projectKey, sourceEnvironmentKey]); // Remove selectedEnvironment and setSelectedEnvironment from dependencies
+        })
+        .catch((error) => {
+          console.error('Error fetching environments:', error);
+        })
+        .finally(() => {
+          setIsLoading(false);
+        });
+    }, 300),
+    [
+      projectKey,
+      sourceEnvironmentKey,
+      selectedEnvironment,
+      setSelectedEnvironment,
+    ],
+  );
 
   useEffect(() => {
-    const filtered = environments.filter(
+    fetchEnvironmentsDebounced(searchQuery);
+  }, [fetchEnvironmentsDebounced, searchQuery]);
+
+  useEffect(() => {
+    const filtered = environments?.filter(
       (env) =>
         env.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
         env.key.toLowerCase().includes(searchQuery.toLowerCase()),
     );
-    setFilteredEnvironments(filtered);
+    setFilteredEnvironments(filtered || []);
```

Comment:
it's not clear to me if we still need to have a reference to `filteredEnvironments`. it seems like the intention was to limit the number of envs we get upfront, so the filtering should now be happening on the backend rather than the frontend? please correct me if i've misunderstood 😅

Reply:
Ooh yea I think you're right! The backend is now doing the filtering :)

Reply:
done